src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs

```diff
@@ -119,14 +119,29 @@ public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationTo
             {
                 using (Logger.LogBlock(FunctionId.SolutionState_ComputeChecksumsAsync, FilePath, cancellationToken))
                 {
-                    // get states by id order to have deterministic checksum.  Limit to the requested set of projects
-                    // if applicable.
+                    // get states by id order to have deterministic checksum.  Limit expensive computation to the
+                    // requested set of projects if applicable.
                     var orderedProjectIds = ChecksumCache.GetOrCreate(ProjectIds, _ => ProjectIds.OrderBy(id => id.Id).ToImmutableArray());
-                    var projectChecksumTasks = orderedProjectIds.Where(id => projectsToInclude == null || projectsToInclude.Contains(id))
-                                                                .Select(id => ProjectStates[id])
-                                                                .Where(s => RemoteSupportedLanguages.IsSupported(s.Language))
-                                                                .Select(s => s.GetChecksumAsync(cancellationToken))
-                                                                .ToArray();
+                    var projectChecksumTasks = orderedProjectIds
+                        .Select(id => (state: ProjectStates[id], mustCompute: projectsToInclude == null || projectsToInclude.Contains(id)))
+                        .Where(t => RemoteSupportedLanguages.IsSupported(t.state.Language))
+                        .Select(async t =>
+                        {
+                            // if it's a project that's specifically in the sync'ed cone, include this checksum so that
+                            // this project definitely syncs over.
+                            if (t.mustCompute)
+                                return await t.state.GetChecksumAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Small allocation nit: you could remove the await here and make the other returns be a Task.FromResult.